### DISCUSSION TITLE
discussion: 有关 Vercel 和 Netlify 的部署问题

### DISCUSSION THREAD

**Discussion by tiwentichat** [nolebase/nolebase#155] [Q&A]
bug： 主页上的 github上浏览对应的链接错了啊大佬们！ 

https://github.com/nolebase/notebook 这个应该是旧的repo名字，新的是nolebase啊，快改一下吧（辛亏我机智hiahiahia

提议：

可以直接部署到vercel或者netlify上，这个方式对新手相对友好（比如俺！，只需要注册github-用gh登陆vercel/netlify, 然后部署

不过第一次部署会失败，要在seting-config里面改一下，

build comand 默认是pnpm build， 需要改为 pnpm docs:build

out directory默认是dist ，需要改为： .vitepress/dist   别忘了是  .vitepress   不能少了那个.

如果大佬们愿意的话可以把vercel方式加进去哇，或者netlify的部署链接放上去

最后，一个请求：

你们笔记的文件夹名 那些小图标是怎样打上去的啊，是输入法的问题吗还是啥啊，我用的讯飞，死活找不到，哭死 T T

救....救....救......

**Comment by nekomeowww** [comment]
> https://github.com/nolebase/notebook 这个应该是旧的repo名字，新的是nolebase啊，快改一下吧（辛亏我机智hiahiahia

厉害！我看看修改一下，可能是我改漏了 🤣

> 可以直接部署到 Vercel 或者 Netlify 上

谢谢你的建议，我也这么觉得，但真的很抱歉！在做了在做了，这需要我们把笔记和主题分离开，然后还有一部分插件和样式正在缓慢迁移到隔壁的 [nolebase/integrations](https://github.com/nolebase/integrations)。

也许我可以先加一个页面来教如何直接删现成的笔记然后上传到 Vercel / Netlify？我看有很多人他们会在 Repo 里面放一个 Deploy on Netlify 和 Deploy on Vercel 的按钮，点一下就能复制 + 部署，不知道这样会不会也可以呢？

> build comand 默认是pnpm build， 需要改为 pnpm docs:build
> out directory默认是dist ，需要改为： .vitepress/dist 别忘了是 .vitepress 不能少了那个.

辛苦你了 🥲 我再改改文档，把很多细节澄清出来。我在 [构建并渲染为可部署的静态页面](https://nolebase.ayaka.io/%E7%AC%94%E8%AE%B0/#%E6%9E%84%E5%BB%BA%E5%B9%B6%E6%B8%B2%E6%9F%93%E4%B8%BA%E5%8F%AF%E9%83%A8%E7%BD%B2%E7%9A%84%E9%9D%99%E6%80%81%E9%A1%B5%E9%9D%A2) 章节给了如何构建的指南，可以给我一些关于这部分的文档应该如何修改的建议吗？

> 你们笔记的文件夹名 那些小图标是怎样打上去的啊，是输入法的问题吗还是啥啊？

这个其实就是 Emoji。我不知道讯飞应该如何打 Emoji，但是也许你也可以通过复制粘贴网站上的 Emoji 来实现类似的操作？比如 [getemoji.com](https://getemoji.com/) 和 [🤣 Emoji表情大全，颜文字百科 💌 - EmojiXD](https://emojixd.com/)

**Comment by tiwentichat** [comment]
> 我看看修改一下，可能是我改漏了

是的，太靠前了，容易忽略，哈哈

> 我可以先加一个页面来教如何直接删现成的笔记然后上传到 Vercel / Netlify？我看有很多人他们会在 Repo 里面放一个 Deploy on Netlify 和 Deploy on Vercel 的按钮，点一下就能复制 + 部署，不知道这样会不会也可以呢？

我作为微微懂一点代码的弱鸡经验：直接甩那个vercel 或者netlify 是最省心的。这样部署完之后，后面我就只用修改内容就行（正在埋头删本地repo你们的笔记 + 研究哪些是和obisidian的插件相关

如果你们俩愿意折腾的话，可以把代码修改的让v+n平台的默认设置能够直接生效，这样是最傻瓜式的。

但是我感觉那样的话好像太麻烦了，似乎没那必要。毕竟能想到自己跑到github上自己折腾个知识库的，最少代码水平都比我强吧，有那两行提示指引下绰绰有余了

> 教如何直接删现成的笔记？

额，我是直接把笔记下面的内容直接删除啊（不过游戏和资源链接正在按图索骥啧啧啧

有啥是不能删的吗

> 我在 [构建并渲染为可部署的静态页面](https://nolebase.ayaka.io/%E7%AC%94%E8%AE%B0/#%E6%9E%84%E5%BB%BA%E5%B9%B6%E6%B8%B2%E6%9F%93%E4%B8%BA%E5%8F%AF%E9%83%A8%E7%BD%B2%E7%9A%84%E9%9D%99%E6%80%81%E9%A1%B5%E9%9D%A2) 章节给了如何构建的指南，可以给我一些关于这部分的文档应该如何修改的建议吗？

我觉得从程序员的角度来说那样的描述已经够用了

我也是看了这个页面的内容后才去config那儿修改了两处。


> 这个其实就是 Emoji。我不知道讯飞应该如何打 Emoji，但是也许你也可以通过复制粘贴网站上的 Emoji 来实现类似的操作？比如 [getemoji.com](https://getemoji.com/) 和 [🤣 Emoji表情大全，颜文字百科 💌 - EmojiXD](https://emojixd.com/)

好嘞！我之前傻了，把svg图往里面贴。。。 刚刚把😼歪嘴猫贴上去了，感谢Nekomeo!

再次感谢两位大佬的项目！ （其实我最喜欢的是右上角的布局调整功能，obisidian我是第一次用，正在研究它的插件到底有多好使

**Comment by nekomeowww** [comment]
> 好嘞！我之前傻了，把svg图往里面贴。。。 刚刚把😼歪嘴猫贴上去了，感谢Nekomeo!

哈哈。看来这部分也得讲讲和科普一下。

另外给你推荐 [Iconify](https://icones.js.org/) 这个开源图标库索引网站，在这个网站里面你可以选择不同的开源图标库集合，然后在本地安装图标库之后直接在文档中使用和上色，包含多色的图标和动画图标。

Nolebase 现在的 `package.json` 中引入的额外的几个 Iconify 图标库，你都可以直接从 [Iconify](https://icones.js.org/) 上面复制下来作为文档内的图标去使用：

https://github.com/nolebase/nolebase/blob/c376563f2456f973fdaefd8f59327d502f28f514/package.json#L18-L20

直接在复制 `@iconify-json/` 后面的字符然后到 [Iconify](https://icones.js.org/) 上搜索就能看到图标集，点进去之后就能开始选自己喜欢的图标。

选定自己喜欢的图标之后，可以点击这个复制图标复制图标的 ID

![image](https://github.com/nolebase/nolebase/assets/11081491/1073f3b5-d5cf-4f0f-89bd-d517c38921b9)

然后在 Markdown 里面直接这么写

```html
<span class="i-<复制进来的图标>" />
```

或者对于上面的截图的例子而言就是

```html
<span class="i-octicon:archive-16" />
```

就可以用了！如果想改颜色的话可以通过行内 HTML 的样式来实现

```html
<span class="i-octicon:archive-16" style="color: red;" />
```

**Comment by tiwentichat** [comment]
> > 好嘞！我之前傻了，把svg图往里面贴。。。 刚刚把😼歪嘴猫贴上去了，感谢Nekomeo!
> 
> 哈哈。看来这部分也得讲讲和科普一下。
> 
> 另外给你推荐 [Iconify](https://icones.js.org/) 这个开源图标库索引网站，在这个网站里面你可以选择不同的开源图标库集合，然后在本地安装图标库之后直接在文档中使用和上色，包含多色的图标和动画图标。
> 
> Nolebase 现在的 `package.json` 中引入的额外的几个 Iconify 图标库，你都可以直接从 [Iconify](https://icones.js.org/) 上面复制下来作为文档内的图标去使用：
> 
> https://github.com/nolebase/nolebase/blob/c376563f2456f973fdaefd8f59327d502f28f514/package.json#L18-L20
> 
> 直接在复制 `@iconify-json/` 后面的字符然后到 [Iconify](https://icones.js.org/) 上搜索就能看到图标集，点进去之后就能开始选自己喜欢的图标。
> 
> 选定自己喜欢的图标之后，可以点击这个复制图标复制图标的 ID
> 
> ![image](https://private-user-images.githubusercontent.com/11081491/289542212-1073f3b5-d5cf-4f0f-89bd-d517c38921b9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIzMDAwMzAsIm5iZiI6MTcwMjI5OTczMCwicGF0aCI6Ii8xMTA4MTQ5MS8yODk1NDIyMTItMTA3M2YzYjUtZDVjZi00ZjBmLTg5YmQtZDUxN2MzODkyMWI5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjExVDEzMDIxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcwMDQyZmMwYzIzZDdlZjQ5ZGJhODE0M2VmMmQxMzdjNTY0NDk4OGJhNzUxMmNmYmY5MTg2ZjBlNDdlYmE2OTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.R9aFiLoLLqqCQnWjv3dMdGsPbHNKOvaB_YH3I1aqlGc)
> 
> 然后在 Markdown 里面直接这么写
> 
> ```
> <span class="i-<复制进来的图标>" />
> ```
> 
> 或者对于上面的截图的例子而言就是
> 
> ```
> <span class="i-octicon:archive-16" />
> ```
> 
> 就可以用了！如果想改颜色的话可以通过行内 HTML 的样式来实现
> 
> ```
> <span class="i-octicon:archive-16" style="color: red;" />
> ```

是的，我就是在这个iconify里面下的svg，然后又尝试贴了那个复制的代码到文件夹名称里，被obi拒绝了说不准我加非法字符，也就是: 

复制代码放在<span>代码里面我模糊会一点点( 虽然也忘的差不多了，正好把你刚才这段当教程!



刚发现还有一个奇怪的，就是在obi的软件里面，打开根目录的index.md， feature 和hero两条目后面的内容是无法修改的。在jetbrains fleet编辑器里面是可以修改的。

这个点不知道是为什么

**Comment by nekomeowww** [comment]
> 是的，我就是在这个iconify里面下的svg

很厉害！很先进！给你夸夸。

> 刚发现还有一个奇怪的，就是在obi的软件里面，打开根目录的index.md， feature 和hero两条目后面的内容是无法修改的。在jetbrains fleet编辑器里面是可以修改的。
> 
> 这个点不知道是为什么

feature 和 hero 所在的这部分由 `---` 分割开的字符叫做 frontmatter，Obsidian 里面是把 frontmatter 当作是键值对来处理的，也就是只支持

```yaml
key: val
key2: val2
```

的形式。而 JetBrain 和 VSCode 都是很通用的编辑器，所以可以用，我不知道 Obsidian 有没有能允许编辑的插件，但也许又是一个新的开发方向了吧，之后如果找不到插件的话也可以想办法自己开发一个 😋

**Comment by nekomeowww** [comment]
> 我作为微微懂一点代码的弱鸡经验：直接甩那个vercel 或者netlify 是最省心的。

辛苦你了！

> 如果你们俩愿意折腾的话，可以把代码修改的让v+n平台的默认设置能够直接生效，这样是最傻瓜式的。

没问题的，我看看。Netlify 可以提供 build 用的配置文件，我可以提供一下，Vercel 的话我得看看。

> 额，我是直接把笔记下面的内容直接删除啊（不过游戏和资源链接正在按图索骥啧啧啧

可以的，理论上 `笔记` 这个目录下的东西删了重新写就行，如果想要直接删掉 `笔记` 的话，还得看看 `scripts/` 目录下的脚本，我印象里是硬编码了搜索 `笔记` 这个目录下面的文件夹和文件来生成侧边栏的，如果你要改名字的话，`scripts/` 目录下也得搜索一下 `笔记` 这两个字然后替换修改一下。

> 有啥是不能删的吗

- `.vitepress/`
- `scripts/`
- `public/`

下面的东西不能随便删，得看看有没有引用才行。唉，真的抱歉，我们还在做迁移，这个时间点上突然有这么可爱的家伙来尝试部署了拿来用，又没有文档可以参考，真的很抱歉。

另外：

- `toc.md` 对应的是部署之后的 `最近更新`，删掉的话这个功能就丢失了，但是肯定不会报错
- `netlify/` 下面的脚本应该是链接跳转用的，可以放心删除。

**Comment by nekomeowww** [comment]
> 正在埋头删本地repo你们的笔记 + 研究哪些是和obisidian的插件相关

在 https://nolebase.ayaka.io/to/b5450a5ec6 这么个隐藏页面里，有讲述我们知识库必须依赖的 Obsidian 插件。

你可以试着全文搜索一下 `dataview`，有这个注解的地方都是依赖了 Obsidian 的 Dataview 的代码块，需要有 Obsidian Dataview 插件才能看到。我印象里应该都在 `视图` 这个目录下。

然后就是 Markdown 里面会有脚注，这些脚注都是 Obsidian 的 Footnote 插件提供的。

**Comment by nekomeowww** [comment]
> 其实我最喜欢的是右上角的布局调整功能，obisidian我是第一次用，正在研究它的插件到底有多好使

谢谢你喜欢，你可以在 https://nolebase-integrations.ayaka.io/pages/zh-CN/integrations/vitepress-plugin-enhanced-readabilities/ 找到它的文档，为了能够在 VitePress 实现部分的 Obsidian 的功能，我也在 https://nolebase-integrations.ayaka.io/pages/zh-CN 开发了一些插件来解决问题，也许可以给你参考参考。

**Comment by nekomeowww** [comment]
链接已经修好了，谢谢你 🥰

**Comment by tiwentichat** [comment]
> 如果找不到插件的话也可以想办法自己开发一个 
哇！厉害！

我目前是双持，fleet作为补充，不过主要还是写内容，所以这个插件不一定那么大众。

PS： 又在iconify遨游了半天，感觉真的是个宝藏啊!

> Netlify 可以提供 build 用的配置文件，我可以提供一下，Vercel 的话我得看看。

v和n的内在逻辑应该差不多，都是部署站点的。

我拿同样的项目两边都部署过，V的界面更清新简洁现代一些，n的图标和界面什么的感觉很老派的样子。

n的好处是部署完后那个xxxx.netlify.app 的链接直接可以访问，不加自己的域名也行；推荐你们迁移好之后生成一个n的站点挂出来，国内访问无障碍

v：清新简洁优雅美观，但：xxxx.vercel.app 是被禁的死死的，想打开是不可能的，这辈子都别想访问...
不过它的使用逻辑十分清晰，而且站点搭建完之后会有一个首页预览，感觉产品经理里面有不懂代码的（或者带入到不懂代码的状态），毕竟V和N的目的就是方便部署+使用啊


> 硬编码了搜索 笔记 这个目录下面的文件夹和文件来生成侧边栏

对，我看到了有笔记之间的关联图示，觉得蛮好的，在看它们是怎样生效的。

- .vitepress/
- scripts/
- public/
- toc.md

我知道是不能删

- netlify 

我还正迟疑来着，那我就大胆删除了~~

>  https://nolebase.ayaka.io/to/b5450a5ec6 

没错，我正在对着它去理解那些插件。在obi里面可以看到这个文档

感觉插件太多， 学不过来了 T T

> 链接已经修好了

✌️ 能帮助到你们最好了

无以为报，送你们一张猫图吧！

![494235839](https://github.com/nolebase/nolebase/assets/125454744/a1e1ef5c-df80-4fbd-8561-f065518faa1f)

赵·赖皮·能吃·虎 & 虞·黑妞·登高望远·姬

**Comment by Lionad-Morotar** [comment]
> 你们笔记的文件夹名 那些小图标是怎样打上去的啊，是输入法的问题吗还是啥啊，我用的讯飞，死活找不到，哭死 T T

如果是 Windows 的话可以简单按下 `windows` + `;` 就能弹出系统自带的 Emoji 搜索器啦~

![image](https://github.com/nolebase/nolebase/assets/26913795/1c6c2ce7-84bd-4a97-918f-fa3d77f692ba)

**Comment by jyuuroku16** [comment]
> bug： 主页上的 github上浏览对应的链接错了啊大佬们！
> 
> https://github.com/nolebase/notebook 这个应该是旧的repo名字，新的是nolebase啊，快改一下吧（辛亏我机智hiahiahia
> 
> 提议：
> 
> 可以直接部署到vercel或者netlify上，这个方式对新手相对友好（比如俺！，只需要注册github-用gh登陆vercel/netlify, 然后部署
> 
> 不过第一次部署会失败，要在seting-config里面改一下，
> 
> build comand 默认是pnpm build， 需要改为 pnpm docs:build
> 
> out directory默认是dist ，需要改为： .vitepress/dist 别忘了是 .vitepress 不能少了那个.
> 
> 如果大佬们愿意的话可以把vercel方式加进去哇，或者netlify的部署链接放上去
> 
> 最后，一个请求：
> 
> 你们笔记的文件夹名 那些小图标是怎样打上去的啊，是输入法的问题吗还是啥啊，我用的讯飞，死活找不到，哭死 T T
> 
> 救....救....救......

补充下，在vercel部署默认使用node18会部署失败，切到node16才部署成功

**Comment by nekomeowww** [comment]
> 补充下，在vercel部署默认使用node18会部署失败，切到node16才部署成功

这么多不兼容的问题吗 😂。
我最近发现 Node 20 也有问题，无法构建，得切换到 Node 21 才行，大家要不要试试看直接切换到 Node 21 呢？

**Comment by jyuuroku16** [comment]
> > 补充下，在vercel部署默认使用node18会部署失败，切到node16才部署成功
> 
> 这么多不兼容的问题吗 😂。 我最近发现 Node 20 也有问题，无法构建，得切换到 Node 21 才行，大家要不要试试看直接切换到 Node 21 呢？

<img width="1008" alt="image" src="https://github.com/nolebase/nolebase/assets/50625641/6a628b12-26b0-4d6f-88c5-9136d469a6f7">
vercel貌似无法切到21

**Comment by nekomeowww** [comment]
> Vercel 貌似无法切到 21

可以尝试通过用 `package.json` 下面的 `engines` 字段的 `node` 子字段来覆盖 Vercel 的 Node.js 版本吗？

文档：

- [https://docs.npmjs.com/cli/v10/configuring-npm/package-json#engines](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#engines)
- [https://vercel.com/docs/functions/runtimes/node-js#version-overrides-in-package.json](https://vercel.com/docs/functions/runtimes/node-js#version-overrides-in-package.json)
